hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/observer/ObservableBatch.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.impl.observer;
+
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Objects;
+
+public class ObservableBatch implements IdentifiedDataSerializable {
```

Comment:
I wonder if it's really a good idea to publish in batches, as the user will not have direct control over how many items are stored in the ringbuffer.  You may have a ringbuffer of size 1000, but each batch could be 5,000 items, increasing the expected memory usage by 5000.

Reply:
Ringbuffer should support batching the publish, but not sure if ReliableTopic supports it.

Reply:
Ok, `Ringbuffer` does have an `addAllAsync` method, which allows for batching, in the sense that it should be more efficient than multiple simple adds (according to the javadoc), but as of yet I don't really see any big issues with using `ReliableTopic`, which would warrant moving over to `Ringbuffer` and potentially reimplement extra functionality currently handled by `ReliableTopic`. Do you?

Maybe as a compromise I could limit the maximum items carried by `ObservableBatch` to something like 10-16? That would limit the possible memory usage.

What do you think?

Reply:
I think limiting batch size will be good to keep memory usage in check. I don't know a good number though, and what will be the size of the ringbuffer by default?

Reply:
The default size for the ringbuffer is 10_000. For the batch size I was thinking to pick a value that would make the Object[] fit on a cache line, but that would be something like 6, which seems too small... Maybe we could go with some "random" round value like 16, until we find a reason to change it. But such a value, together with the 10_000 default size of the ringbuffer suggests that maybe we could remove this batching entirely. Do you see any extra arguments for either approach?

Reply:
Without batching it means we will need one op per every item, which will be slow. Using one item per slot, but being able to add multiple add a time would be the most optimal option.

Reply:
if the ringbuffer has batch api, it should be easy to add it for `ReliableTopic`. As a non-public methods, they don't like new methods unless requested by a user.

Reply:
Should I ask for it to be added or should I submit a PR for it? What to do in the meantime, put in some made up limitation for the buffer sizes as a temporary improvement?

Reply:
I think we can just add one by one for now, and then add the batching after this PR is merged.